flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/AbstractLeaderElectionService.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.leaderelection;
+
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.UUID;
+
+/**
+ * Abstract class for leader election service based on distributed coordination system(e.g. Zookeeper, Kubernetes, etc.).
+ */
+public abstract class AbstractLeaderElectionService implements LeaderElectionService {
+
+	protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+	protected final Object lock = new Object();
+
+	/** The leader contender which applies for leadership. */
+	protected volatile LeaderContender leaderContender;
+
+	private volatile UUID issuedLeaderSessionID;
+
+	protected volatile UUID confirmedLeaderSessionID;
+
+	protected volatile String confirmedLeaderAddress;
+
+	protected volatile boolean running;
+
+	protected AbstractLeaderElectionService() {
+		leaderContender = null;
+
+		issuedLeaderSessionID = null;
+		confirmedLeaderSessionID = null;
+		confirmedLeaderAddress = null;
+
+		running = false;
+	}
+
+	@Override
+	public final void start(LeaderContender contender) throws Exception {
+		Preconditions.checkNotNull(contender, "Contender must not be null.");
+		Preconditions.checkState(leaderContender == null, "Contender was already set.");
+
+		logger.info("Starting LeaderElectionService {}.", this);
+
+		synchronized (lock) {
+			leaderContender = contender;
+			running = true;
+			internalStart(contender);
+		}
+	}
+
+	@Override
+	public final void stop() throws Exception {
+		synchronized (lock) {
+			if (!running) {
+				return;
+			}
+			running = false;
+			clearConfirmedLeaderInformation();
+		}
+
+		logger.info("Stopping LeaderElectionService {}.", this);
+
+		internalStop();
+	}
```

Comment:
Actually, there's no valid harm to also put `internalStop` in the `synchronized` block. Since the old behavior in `ZookeeperLeaderElectionService` is like this, I keep them in the `AbstractLeaderElectionService`. Just like you said, it is a little hard for the maintenance. I agree that we could move the `internalStop` into the `synchronized` block.